### ISSUE TITLE
JS 关于jQuery

### ISSUE THREAD

**Issue by plainnany** [plainnany/blog#27]
##  jQuery 能做什么？
1. 选择某个网页元素，然后对其进行某种操作，如$(document) 、$('#myId') 、$('div.myClass') 
2. 改变结果集，jQuery中提供各种强大的过滤器，对结果集进行筛选，缩小选择结果。如
```
$('div').eq(5); //选择第6个div元素
$('div').first(); //选择第1个div元素
$('div').not('.myClass'); //选择class不等于myClass的div元素
```
3. 链式操作，jQuery中最终选中网页元素以后，可以对它进行一系列操作，并且所有操作可以连接在一起，以链条的形式写出来，比如：
```
$('div').find('h3').eq(2).html('Hello');
```
4. 操作网页元素，最常见的需求是取得它们的值，或者对它们进行赋值。
5. 元素的操作：复制、删除和创建，复制元素使用.clone()，删除元素使用.remove()和.detach()。两者的区别在于，前者不保留被删除元素的事件，后者保留，有利于重新插入文档时使用，清空元素内容（但是不删除该元素）使用.empty()。
6. 事件操作，把事件直接绑定在网页元素上，
```
$('p').click(function(){
　　　alert('Hello');
});
```
7. 浏览器兼容性较高。
等
##  jQuery 对象和 DOM 原生对象有什么区别？如何转化？
jQuery的对象是具有jQuery定义的一些属性和方法的，且它是一个伪数组，
dom对象只用于dom原生的属性和方法，可以是单独的dom节点个体，也可以是一个集合
```
// <div class="box"></box>
$('div')[0]  // 原生dom对象
$('.box')[0]  // 原生dom对象
$(.box).eq(0)  // 只有一个节点的类数组
```
## jQuery中如何绑定事件？bind、unbind、delegate、live、on、off都有什么作用？推荐使用哪种？使用on绑定事件使用事件代理的写法？
jQuery中一般用on绑定事件，$(XX).on( events [,selector ] [,data ], handler(eventObject) )，这是1.7版本后统一的写法。在这之前，各种事件都有自己单独的绑定语句，比如click，有$(xx).click( [eventData ], handler )。
1. bind() / unbind()也是一种绑定/解绑事件的方法，但是已经被弃用了。
2. delegate()，用来执行事件委托，语法是：$(XX).delegate( selector, eventType, eventData, handler )
3. live()，也是一种事件委托的方式，语法是$(xx).live( events [, data ], handler )，其中第二的参数可以传入selector，live给document绑定一个事件，当有事件触发时，判定事件源是否匹配selector，是的话则执行handler里的语句。
4. off对应于on，用来解绑事件:.off( events [, selector ] )，不传入参数时会移除对象身上所有的事件。

使用on来实现事件代理
```
// <ul class="list"><ul>
$('.list').on('click', 'li' , function(){
  console.log('hi')
})
```
## jQuery 如何展示/隐藏元素？
show()，hide()，fadeIn()，fadeOut()，toggle()，fadeToggle()
```
<div class="hide"></div>
<div class="show"></div>
<div class="content">content</div>

$('hide').on('click', function(){
  $('.content').hide()     // $('.content').fadeOut()
})
$('hide').on('click', function(){
  $('.content').show()   // $('.content').fadeIn()
})
```
##  jQuery 动画如何使用？
用animate()这个API
语法是.animate( properties [, duration ] [, easing ] [, complete ] )
1. Animation Properties and Values(动画属性和值)
所有用于动画的属性必须是数字的，除非另有说明；这些属性如果不是数字的将不能使用基本的jQuery功能。（例如，width, height或者left可以执行动画，但是background-color不能，除非使用jQuery.Color()插件。）属性值的单位像素（px）,除非另有说明。单位em 和 %需要指定使用。
除了样式属性， 一些非样式的属性，如scrollTop 和 scrollLeft，以及自定义属性，也可应用于动画。
CSS简写属性（例如font, background, border）没有得到充分的支持。例如， 如果你想边框宽度做动画呈现，至少边框样式和边框宽度必须预先设定不为"auto" 。或者，如果你想字体大小做动画， 你可以使用fontSize的或相当于CSS的 'font-size' ，而不是简单的'font'。
除了定义数值，每个属性能使用'show', 'hide', 和 'toggle'。这些快捷方式允许定制隐藏和显示动画用来控制元素的显示或隐藏。为了使用jQuery内置的切换状态跟踪，'toggle'关键字必须在动画开始前给定属性值。（愚人码头注：jquery中有两个.toggle()方法，一个是显示或隐藏匹配元素。另一个是绑定两个或多个处理程序绑定到匹配的元素，用来执行在交替的点击。哪一个被调用取决于传递的参数的设置。所以这里'toggle'关键字必须在动画开始前给定属性值，以确认调用哪个'toggle'方法。）
动画属性也可以是一个相对值。如果提供一个以+= 或 -=开始的值，那么目标值就是以这个属性的当前值加上或者减去给定的数字来计算的。
2. Duration（持续时间）
持续时间是以毫秒为单位的；值越大表示动画执行的越慢，不是越快。还可以提供'fast' 和 'slow'字符串，分别表示持续时间为200 和 600毫秒。
注意: 不同于动画
3. Complete Function（完成函数）
如果提供complete回调函数将在动画完成后被执行一次。这对于按顺序执行一系列不同动画时，特别有用。这个回调函数不设置任何参数，但是this会被传递到正在执行动画的DOM元素上，如果多个元素一起做动画效果，值得注意的是这个回调函数在每个匹配元素上执行一次，不是这个动画作为一个整体。

[点击查看效果](http://js.jirengu.com/ceqos/2/edit)

## 如何设置和获取元素内部 HTML 内容？如何设置和获取元素内部文本？
```
$(xxx).html(); //不传参，获取html内容
$(xxx).html('<li>111</li>'); //传参，设置元素内部html内容为所传参数
$(xxx).text();  //不传参，获取元素内部的文字信息
$(xxx).text('hi');// 传参，设置元素内部text内容为所传参数
```
## 如何设置和获取表单用户输入或者选择的内容？如何设置和获取元素属性？
获取表单输入内容.val()，
设置表单输入内容，val('hi')
获取元素属性，attr(attribute)
设置元素属性，attr(attribute,attributeName)
##  使用 jQuery实现如下效果
[demo1](http://js.jirengu.com/fuco)
[demo2 添加](http://js.jirengu.com/yozid)
[demo3 tab切换](http://js.jirengu.com/kome)
[demo4 jQuery动画实现tab](http://js.jirengu.com/cowar)
